File: crates/metatorio-egui/src/factorio/format.rs
const LARGE_UNITS: [&str; 11] = ["", "k", "M", "G", "T", "P", "E", "Z", "Y", "R", "Q"];

pub fn signed_compact_number(num: f64) -> String {
    if num.is_sign_negative() {
        format!("-{}", compact_number(-num))
    } else {
        format!("+{}", compact_number(num))
    }
}

pub fn compact_number(num: f64) -> String {
    let abs_num = num.abs();

    match abs_num {
        n if n < 1e-15 => String::from("0"),
        n if n < 1e-9 => format_with_unit(num * 1e12, "p"),
        n if n < 0.01 => format_with_unit(num * 1e6, "μ"),
        n => {
            let mut unit_idx = 0;
            let mut n = n;
            if n > 10000.0 {
                while n > 1000.0 && unit_idx < LARGE_UNITS.len() - 1 {
                    unit_idx += 1;
                    n /= 1000.0;
                }
            }
            format_with_unit(n * num.signum(), LARGE_UNITS[unit_idx])
        }
    }
}

fn format_with_unit(value: f64, unit: &str) -> String {
    let abs_value = value.abs();
    if unit.is_empty() {
        if abs_value < 10.0 {
            // 对于小于10的值，最多保留2位小数
            let formatted = format!("{:.2}", value);
            format!(
                "{}{}",
                formatted.trim_end_matches('0').trim_end_matches('.'),
                unit
            )
        } else if abs_value < 1000.0 {
            // 对于大于10的值，最多保留1位小数
            let formatted = format!("{:.1}", value);
            format!(
                "{}{}",
                formatted.trim_end_matches('0').trim_end_matches('.'),
                unit
            )
        } else {
            // 对于大于等于1000的值，取整
            format!("{}{}", value.round(), unit)
        }
    } else if abs_value < 10.0 {
        // 对于小于10的值，最多保留2位小数
        let formatted = format!("{:.2}", value);
        format!(
            "{}{}",
            formatted.trim_end_matches('0').trim_end_matches('.'),
            unit
        )
    } else if abs_value < 100.0 {
        // 对于10-100的值，最多保留1位小数
        let formatted = format!("{:.1}", value);
        format!(
            "{}{}",
            formatted.trim_end_matches('0').trim_end_matches('.'),
            unit
        )
    } else {
        // 对于大于等于100的值，取整
        format!("{}{}", value.round(), unit)
    }
}

pub fn parse_number(n: &str) -> Option<f64> {
    let re = regex::Regex::new(r"^-?[\d.e-]+[kMGTPEZYRQμp]?$").ok()?;
    if re.is_match(n) {
        let multiplier = match n.chars().next_back() {
            Some('p') => 0.000_000_000_001,
            Some('μ') => 0.000_001,
            Some('k') => 1_000.0,
            Some('M') => 1_000_000.0,
            Some('G') => 1_000_000_000.0,
            Some('T') => 1_000_000_000_000.0,
            Some('P') => 1_000_000_000_000_000.0,
            Some('E') => 1_000_000_000_000_000_000.0,
            Some('Z') => 1_000_000_000_000_000_000_000.0,
            Some('Y') => 1_000_000_000_000_000_000_000_000.0,
            Some('R') => 1_000_000_000_000_000_000_000_000_000.0,
            Some('Q') => 1_000_000_000_000_000_000_000_000_000_000.0,
            _ => 1.0,
        };
        let numeric_value: f64 = n
            .trim_end_matches(|c: char| !c.is_ascii_digit())
            .parse()
            .ok()?;
        Some(numeric_value * multiplier)
    } else {
        None
    }
}

pub fn parse_energy(n: &str) -> Option<f64> {
    let re = regex::Regex::new(r"^-?[\d.e-]+[kMGTPEZYRQμ]?[JW]$").ok()?;
    if re.is_match(n) {
        let mut multiplier = match n.chars().rev().nth(1) {
            Some('k') => 1_000.0,
            Some('M') => 1_000_000.0,
            Some('G') => 1_000_000_000.0,
            Some('T') => 1_000_000_000_000.0,
            Some('P') => 1_000_000_000_000_000.0,
            Some('E') => 1_000_000_000_000_000_000.0,
            Some('Z') => 1_000_000_000_000_000_000_000.0,
            Some('Y') => 1_000_000_000_000_000_000_000_000.0,
            Some('R') => 1_000_000_000_000_000_000_000_000_000.0,
            Some('Q') => 1_000_000_000_000_000_000_000_000_000_000.0,
            _ => 1.0,
        };
        let dimension_char = n.chars().last();
        if let Some('W') = dimension_char {
            multiplier /= 60.0
        }
        let numeric_value: f64 = n
            .trim_end_matches(|c: char| !c.is_ascii_digit())
            .parse()
            .ok()?;
        Some(numeric_value * multiplier)
    } else {
        None
    }
}

pub fn drag_value<T>(val: &mut T) -> egui::DragValue<'_>
where
    T: egui::emath::Numeric,
{
    egui::DragValue::new(val)
        .custom_parser(parse_number)
        .custom_formatter(|n, _| compact_number(n))
        .update_while_editing(false)
}

pub fn drag_watt<T>(val: &mut T) -> egui::DragValue<'_>
where
    T: egui::emath::Numeric,
{
    egui::DragValue::new(val)
        .suffix("W")
        .custom_parser(|s| parse_energy(s).map(|x| x * 60.0))
        .custom_formatter(|n, _| compact_number(n))
        .update_while_editing(false)
}

#[test]
fn test_compact_format() {
    dbg!(compact_number(1.1));
    dbg!(compact_number(114514.0));
    dbg!(compact_number(1919810.1));
    dbg!(compact_number(123456789.1));
    dbg!(compact_number(0.00011));
}

File: metatorio-app/src/lib/format.ts
// 数量格式化（复刻原始 egui factorio/format.rs 的 compact_number）：
// - 极小值用 p（1e-12）/ μ（1e-6）
// - 大数用 k/M/G/T/P/E/Z/Y/R/Q 前缀
// - 无前缀时按数量级决定小数位（<10 两位、<1000 一位、更大取整）

const LARGE_UNITS = ["", "k", "M", "G", "T", "P", "E", "Z", "Y", "R", "Q"];

function formatWithUnit(value: number, unit: string): string {
  const abs = Math.abs(value);
  const trim = (text: string) => text.replace(/0+$/, "").replace(/\.$/, "");
  let formatted: string;
  if (abs < 10) {
    formatted = trim(value.toFixed(2));
  } else if (abs < 100) {
    formatted = trim(value.toFixed(1));
  } else {
    formatted = String(Math.round(value));
  }
  return `${formatted}${unit}`;
}

/** 数量 → 紧凑文本（复刻 egui compact_number）。 */
export function compactNumber(num: number): string {
  const abs = Math.abs(num);
  if (abs < 1e-15) return "0";
  if (abs < 1e-9) return formatWithUnit(num * 1e12, "p");
  if (abs < 0.01) return formatWithUnit(num * 1e6, "μ");
  let n = abs;
  let unitIdx = 0;
  if (n > 10000) {
    while (n > 1000 && unitIdx < LARGE_UNITS.length - 1) {
      unitIdx += 1;
      n /= 1000;
    }
  }
  return formatWithUnit(n * Math.sign(num), LARGE_UNITS[unitIdx]);
}

/** 带符号数量（正数加 + 前缀）。 */
export function signedCompactNumber(num: number): string {
  return num < 0 ? `-${compactNumber(-num)}` : `+${compactNumber(num)}`;
}

/** 解析紧凑数量文本（p/μ/k/M/G/T…），失败返回 null。 */
export function parseCompactNumber(text: string): number | null {
  const match = /^([+-]?\d*\.?\d+)\s*([pμkMGTPEZYRQ])?$/i.exec(text.trim());
  if (!match) return null;
  const value = Number(match[1]);
  if (!Number.isFinite(value)) return null;
  const unit = (match[2] ?? "").toLowerCase();
  const multipliers: Record<string, number> = {
    p: 1e-12,
    μ: 1e-6,
    k: 1e3,
    m: 1e6,
    g: 1e9,
    t: 1e12,
    e: 1e18,
    z: 1e21,
    y: 1e24,
    r: 1e27,
    q: 1e30,
  };
  return value * (multipliers[unit] ?? 1);
}
